### PR TITLE
feat(treesitter): use upstream format for injection queries

### PR DESCRIPTION
**Pull request by lewis6991**
## Problem

The way we specify injections doesn't align with other Treesitter implementations.

## Solution

Implement injections as documented here:
https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection

This aligns with how Helix specifies injections: https://github.com/helix-editor/helix/tree/master/runtime/queries

Note this assumes `exclude_children` by default and `injection.include-children` must be set to match Nvim's old behaviour.

A part of: #22495

### CODE REVIEW

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 1]
```diff
@@ -365,14 +365,166 @@ local function get_range_from_metadata(node, id, metadata)
   return { node:range() }
 end
 
+---@private
+---@param node TSNode
+---@param id integer
+---@param metadata TSMetadata
+---@return Range4[]
+local function get_node_ranges(node, id, metadata, include_children)
```
Comment (clason) [inline runtime/lua/vim/treesitter/languagetree.lua:373]: We already have `vim.treesitter.get_node_range`-- can this be consolidated?

(May need a -- much needed -- cleanup of the node_range interface, so maybe defer and add a `TODO` here if it's out of scope?)

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 2]
```diff
@@ -365,14 +365,166 @@ local function get_range_from_metadata(node, id, metadata)
...
+
+  table.insert(t[tree_index][lang][pattern].regions, ranges)
+end
+
+---@private
+---@param node TSNode
+---@param source integer|string
+---@param metadata table
+---@return string
+local function get_node_text(node, source, metadata)
```
Comment (clason) [inline runtime/lua/vim/treesitter/languagetree.lua:433]: same -- we already have a `vim.treesitter.get_node_text()`

**Review by clason — COMMENTED** [review]
Can you summarize how backward compatibility operates (i.e., what decides whether the new or old strategy is used)? Is it worth adding notes in the code on what can be removed after nvim-treesitter updates all its queries?

File: runtime/lua/vim/treesitter/languagetree.lua (hunk 1 above)
Comment (lewis6991) [reply runtime/lua/vim/treesitter/languagetree.lua:373]: Definitely out of scope. This was just to break the logic up.

File: runtime/lua/vim/treesitter/languagetree.lua (hunk 2 above)
Comment (lewis6991) [reply runtime/lua/vim/treesitter/languagetree.lua:433]: This is just to shut up the LSP and avoid hacky annotations.

**Comment by lewis6991** [comment]
The new strategy is used first on a match, if they doesn't return an injection then the old one is tried. This should be pretty clear in `_get_injections()`

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 3]
```diff
@@ -388,75 +540,11 @@ function LanguageTree:_get_injections()
...
-            lang = name
-          end
-
-          if #ranges == 0 then
-            table.insert(ranges, get_range_from_metadata(node, id, metadata))
-          end
-        end
+      local lang, combined, ranges = self:_get_injection(match, metadata)
+      if not lang then
+        lang, combined, ranges = self:_get_injection_deprecated(match, metadata)
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/languagetree.lua:545]: Right here

**Comment by clason** [comment]
> The new strategy is used first on a match, if that doesn't return an injection then the old one is tried. This should be pretty clear in `_get_injections()`

Maybe still worth a note, at least in the commit/PR?

File: runtime/lua/vim/treesitter/languagetree.lua (hunk 2 above)
Comment (clason) [reply runtime/lua/vim/treesitter/languagetree.lua:433]: worth a TODO comment then, otherwise you'll get hop-on comments (like this one)

File: runtime/lua/vim/treesitter/languagetree.lua (hunk 1 above)
Comment (clason) [reply runtime/lua/vim/treesitter/languagetree.lua:373]: Could you add a TODO then?

**Comment by lewis6991** [comment]
> And to be clear: the goal is that "upstream format" is the format, going forward? Our old format is just being kept around for backward compatibility and could be removed in due time?

Yes. I was never a fan of our format anyway. It is a bit too loose and the logic for handling it is hard to understand.

**Comment by clason** [comment]
I fully agree; would be good to make this explicit (PR/commit message, docs, `TODO: remove` comment)!

**Comment by justinmk** [comment]
> I fully agree; would be good to make this explicit (PR/commit message, docs, `TODO: remove` comment)!

Added "remove support for old query format" as a post-release task to https://github.com/neovim/neovim/issues/20434 . But +1 for a formal mechanism, so that list could become simply: "Remove all `TODO: remove` (or whatever) things".

File: runtime/doc/treesitter.txt [hunk 4]
```diff
@@ -477,7 +477,53 @@ individual query pattern manually by setting its `"priority"` metadata
...
+associated with patterns:
+
+    • `injection.language` - can be used to hard-code the name of a specific
+    language.
+    • `injection.combined` - indicates that all of the matching nodes in the
+    tree should have their content parsed as one nested document.
+    • `injection.include-children` - indicates that the @injection.content
+    node’s entire text should be re-parsed, including the text of its child
+    nodes. By default, child nodes’ text will be excluded from the injected
+    document.
```
Comment (clason) [inline runtime/doc/treesitter.txt:525]:
```suggestion
    • `injection.include-children` - indicates that the `@injection.content`
    node's entire text should be re-parsed, including the text of its child
    nodes. By default, child nodes' text will be excluded from the injected
    document.
```

**Review by clason — APPROVED** [review]

**Comment by lewis6991** [comment]
I'll try to add some simple tests before merging.

File: runtime/doc/news.txt [hunk 5]
```diff
@@ -189,6 +189,10 @@ The following new APIs or features were added.
   disabled by default and can be enabled by setting the
   `workspace.didChangeWatchedFiles.dynamicRegistration=true` capability.
 
+• Treesitter injection queries now use the format described at
+  https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection .
+  Support for the previous format will be removed in a future release.
```
Comment (justinmk) [inline runtime/doc/news.txt:194]: This is a strong signal that this is the right change. When you can just point to upstream spec and say "we support [this]" instead of "[big paragraph about differences]", that is a gain for both developers and users.

An early, analogous case was in 2014 when we debated a variation of msgpack-rpc. https://github.com/neovim/neovim/pull/1121

File: runtime/doc/treesitter.txt [hunk 6]
```diff
@@ -441,7 +441,53 @@ individual query pattern manually by setting its `"priority"` metadata
...
+    • JavaScript files, which contain regular expression syntax within regex
+      literals
+    • Ruby, which can contain snippets of code inside of heredoc literals,
+      where the heredoc delimiter often indicates the language
+    • Lua, which can contain snippets of Vimscript inside |vim.cmd()| calls.
+    • Vimscript, which can contain snippets of Lua inside |:lua-heredoc|
+      blocks.
+
+All of these examples can be modeled in terms of a parent syntax tree and one
+or more injected syntax trees, which reside inside of certain nodes in the
```
Comment (justinmk) [inline runtime/doc/treesitter.txt:469]: This overlaps with the `@defgroup` in `runtime/lua/vim/treesitter/languagetree.lua` , would it make sense to move this info there?

Comment (lewis6991) [reply runtime/doc/treesitter.txt:469]: Sure, but not sure how? Can you push a commit?

**Comment by lewis6991** [comment]
Keen to get this in. Doc changes can be done in a follow up.

**Comment by justinmk** [comment]
For reference: upstream is considering changing the order of query resolution to match Nvim's original "last wins" approach: https://github.com/nvim-treesitter/nvim-treesitter/issues/3944#issuecomment-1458782497 

> query precedence order - This is fairly minor, but right now, the tree-sitter-highlight crate prefers matches listed earlier in the query, so that you need to list the most specific patterns first. I noticed that Neovim uses the opposite convention, and I think this may be slightly more intuitive, because it matches the behavior of the "cascade" in CSS. Also, in Zed, we also found it slightly easier to implement with the more specific matches occurring later (like Neovim).

**Comment by clason** [comment]
We'll see if they actually change anything, or if this is just "hindsight is 20/20"...
